runtime/common/Trace.h

```diff
@@ -0,0 +1,50 @@
+/****************************************************************-*- C++ -*-****
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#pragma once
+
+#include <string>
+#include <vector>
+
+namespace cudaq {
+
+/// @brief A trace is a circuit representation of the executed computation, as
+/// seen by the execution manager. (Here, a circuit is represented as a list
+/// of instructions on qudits). Since the execution manager cannot "see" control
+/// flow, the trace of a kernel with control flow represents a single execution
+/// path, and thus two calls to the same kernel might produce traces.
+class Trace {
+public:
+  struct Instruction {
+    std::string name;
+    std::vector<double> params;
+    std::vector<std::size_t> controls;
+    std::vector<std::size_t> targets;
```

Comment:
do you think we'll want to retain `QuditInfo` here? So the `Instruction` will also know the number of qudit levels?

Reply:
The two analyses that we do on top of it are agnostic to the number of levels---both only care about unique identifier. I chose this way mainly because it is how `estimate_resources` currently work. It should not be hard to modify to retain it.